### PR TITLE
Add back support for python 3.9 for numpy 2 compatibility woes

### PR DESCRIPTION
**Pull request by hmaarrfk**
It seems highly probable that numpy 2 will release a build for Python 3.9
@rgommers had asked us to add an upper bound to our releases in Jan but it seems that we didn't take action (understanding how pip solves packages is unclear to me even).

Instead of releasing 0.22.1 for python 3.9 with metadata updates, i'm curious to see if we can release a new version with python 3.9.

The headache of getting CIs working is what I'm hoping to avoid with this startegy

xref: https://mail.python.org/archives/list/numpy-discussion@python.org/thread/AHTATJKGUEOILBNUI5IGGZPXJ5FXIRAU/
xref: https://github.com/scikit-image/scikit-image/issues/7282#issuecomment-1885659412

- [ ] Make a PR to scikit-image-build

## Checklist

- A descriptive but concise pull request title
- [Docstrings for all functions](https://github.com/numpy/numpydoc/blob/main/doc/example.py)
- [Unit tests](https://scikit-image.org/docs/dev/development/contribute.html#testing)
- A gallery example in `./doc/examples` for new features
- [Contribution guide](https://scikit-image.org/docs/dev/development/contribute.html) is followed

## Release note

For maintainers and optionally contributors, please refer to the [instructions](https://scikit-image.org/docs/stable/development/contribute.html#documenting-changes) on how to document this PR for the release notes.

```release-note
Add support back for Python 3.9 to enhance compatibility with Numpy 2.
```

### CODE REVIEW

**Comment by hmaarrfk** [comment]
i expected some tests to fail, but I did not expect them to catastrophically fail like this on CPython 3.9

File: skimage/transform/tests/test_geometric.py [hunk 1]
```diff
@@ -747,10 +747,11 @@ def test_inverse_all_transforms(tform):
     assert_almost_equal(tform.inverse.inverse(SRC), tform(SRC))
     # Test addition with inverse, not implemented for all
     if not isinstance(
-        tform,
-        EssentialMatrixTransform
-        | FundamentalMatrixTransform
```
Comment (hmaarrfk) [inline skimage/transform/tests/test_geometric.py:752]: man, this syntax is really strange...

**Comment by hmaarrfk** [comment]
I think this is ready again for review.

While I was very hopeful that NEP29/SPEC0 could be adhered to, I feel like we might need to rethink the general strategy of:

"Does numpy continue support the longest"
or
"Is numpy the first to drop support for older pythons"

Given the likely release of Numpy2 for Python 3.9, I think we should continue to support it, at least until the major bugs have been caught with it.

**Comment by lagru** [comment]
But I'd probably prefer a patch release 0.23.3 with the relevant commits cherry-picked. I'm not sure what you are suggesting as a new version otherwise in 
> i'm curious to see if we can release a new version with python 3.9.

**Comment by hmaarrfk** [comment]
There are two options in my mind:

1. Going back to 0.22.1 + python 3.9
2. Releasing 0.23.3 + python 3.9

I'm providing the option for 0.23.3 since I think that getting the CIs working for 0.22.X will be hard. The work on the LTS still haunts me.

File: skimage/transform/tests/test_geometric.py (hunk 1 above)
Comment (hmaarrfk) [reply skimage/transform/tests/test_geometric.py:752]: FYI it seems that this was added by some autofromatting code

https://github.com/scikit-image/scikit-image/commit/d93ce536974116ba8c7f890f6487aedef2bb3b37

@jarrodmillman is there a way for me to disable this kind of syntax reformatting?

**Comment by hmaarrfk** [comment]
The CIs are green. Any interest in moving on this?

**Comment by stefanv** [comment]
I'd like to have @jarrodmillman weigh in.

**Comment by hmaarrfk** [comment]
it would be good to get a move on this if there is desire.

We are quite good at releasing quality packages, I think this will help many build confidence in the python ecosystem if numpy 2 doesn't break their pipelines.

**Review by mkcor — APPROVED** [review]
I'm very much in favour of this because:

> We are quite good at releasing quality packages, I think this will help many build
> confidence in the python ecosystem if numpy 2 doesn't break their pipelines.

:+1: 

But @jarrodmillman is the expert! :bow: 

For my own understanding, @hmaarrfk, what does

> "Does numpy continue support the longest"

mean? I get the question

> "Is numpy the first to drop support for older pythons"

and it would be good to know if that's an absolute "yes."

**Comment by hmaarrfk** [comment]
> "Does numpy continue support the longest"

should be

> "Does numpy continue support python's the longest"

**Comment by jarrodmillman** [comment]
+1 from me. I will take a closer look at this today (at the Scientific Python Developer Summit!) and get something out soon.

**Review by jarrodmillman — APPROVED** [review]

File: skimage/transform/tests/test_geometric.py (hunk 1 above)
Comment (jarrodmillman) [reply skimage/transform/tests/test_geometric.py:752]: Yes, I will look into it.
- https://docs.astral.sh/ruff/rules/non-pep604-isinstance/
- https://github.com/astral-sh/ruff/issues/7871

Comment (jarrodmillman) [reply skimage/transform/tests/test_geometric.py:752]: See #7430.

**Comment by mkcor** [comment]
Does anyone know why we have
https://github.com/scikit-image/scikit-image/blob/c206205b2305c91f688f47656d77a66ca1d4ed7b/pyproject.toml#L65
? Just noticed:
```bash
$ pip install -r requirements.txt
Ignoring tomli: markers 'python_version < "3.11"' don't match your environment
(...)
```
while setting up my dev env in Python 3.11.3!

**Comment by neutrinoceros** [comment]
@mkcor tomli was added to the standard library (as `tomllib`) in Python 3.11.0, so it's normally not used as a dependency if you're running on 3.11 or newer.

**Comment by mkcor** [comment]
> @mkcor tomli was added to the standard library (as `tomllib`) in Python 3.11.0, so it's normally not used as a dependency if you're running on 3.11 or newer.

Thank you, @neutrinoceros, good to know! I didn't know how to interpret that `tomli` only supported Python from 3.7 through 3.10 (https://pypi.org/project/tomli/).

**Pull request merged**